### PR TITLE
Added liveness for elasticsearch

### PR DESCRIPTION
**Pull request by sushma1118**
*Created a folder liveness inside apps/elasticsearch inside which following scripts are added.
      * Added elasticsearch_liveness.yml file
      * Added run_litmus_test.yml
      * Added test.yml 
      * Added vars.yml 
Used the image openebs/tests-custom-client:latest for python script.
Referred custom-liveness of test-tools/custom-client

@giri @nsathyaseelan @shashank855 

Some files for reference-
Liveness pod is runnig-
```
# kubectl logs elasticsearch-liveness-xjpr6 -n app-esearch-ns -f
pass True
Liveness Running
Liveness Running
Liveness Running
Liveness Running
Liveness Running
Liveness Running
Liveness Running
Liveness Running
Liveness Running
Liveness Running
Liveness Running
Liveness Running
Liveness Running
Liveness Running
Liveness Running
Liveness Running
Liveness Running
Liveness Running
```

when the application pod is deleted liveness is failed

```
kubectl logs elasticsearch-liveness-xjpr6 -n app-esearch-ns -f
pass True
Liveness Running
Liveness Running
Liveness Running
Liveness Running
Liveness Running
Liveness Running
Liveness Running
Liveness Running
Liveness Running
Liveness Running
Liveness Running
Liveness Running
Liveness Running
Liveness Running
Liveness Running
Liveness Running
Liveness Running
Liveness Running
Liveness Running
Liveness Running
Liveness Failed
Retrying to establish connection
Liveness Running
Liveness Running
Liveness Running
Liveness Running
Liveness Running
Liveness Running
Liveness Running
```

Litmus pod log-

```
kubectl logs litmus-elasticsearch-liveness-cb6fp-4sbdv -n litmus f
Error from server (BadRequest): container f is not valid for pod litmus-elasticsearch-liveness-cb6fp-4sbdv
[root@master-1554817485 liveness]# kubectl logs litmus-elasticsearch-liveness-cb6fp-4sbdv -n litmus -f
No config file found; using defaults
/etc/ansible/hosts did not meet host_list requirements, check plugin documentation if this is unexpected
/etc/ansible/hosts did not meet script requirements, check plugin documentation if this is unexpected

PLAY [localhost] ***************************************************************

TASK [Gathering Facts] *********************************************************
ok: [127.0.0.1]

TASK [Record test instance/run ID] *********************************************
skipping: [127.0.0.1] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [Construct testname appended with runID] **********************************
skipping: [127.0.0.1] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [include_tasks] ***********************************************************
included: /utils/fcm/update_litmus_result_resource.yml for 127.0.0.1

TASK [Generate the litmus result CR to reflect SOT (Start of Test)] ************
changed: [127.0.0.1] => {"changed": true, "checksum": "dbe359b043a5ba021072fa35da1a4a3af333742e", "dest": "./litmus-result.yaml", "gid": 0, "group": "root", "md5sum": "25e2ed95c837e380662770ee51efffff", "mode": "0644", "owner": "root", "size": 423, "src": "/root/.ansible/tmp/ansible-tmp-1558358099.92-88751687323931/source", "state": "file", "uid": 0}

TASK [Analyze the cr yaml] *****************************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "cat litmus-result.yaml", "delta": "0:00:01.109493", "end": "2019-05-20 13:15:02.371371", "rc": 0, "start": "2019-05-20 13:15:01.261878", "stderr": "", "stderr_lines": [], "stdout": "---\napiVersion: litmus.io/v1alpha1\nkind: LitmusResult\nmetadata:\n\n  # name of the litmus testcase\n  name: elasticsearch-liveness \nspec:\n\n  # holds information on the testcase\n  testMetadata:\n    app:  \n    chaostype:  \n\n  # holds the state of testcase,  manually updated by json merge patch\n  # result is the useful value today, but anticipate phase use in future \n  testStatus: \n    phase: in-progress  \n    result: none ", "stdout_lines": ["---", "apiVersion: litmus.io/v1alpha1", "kind: LitmusResult", "metadata:", "", "  # name of the litmus testcase", "  name: elasticsearch-liveness ", "spec:", "", "  # holds information on the testcase", "  testMetadata:", "    app:  ", "    chaostype:  ", "", "  # holds the state of testcase,  manually updated by json merge patch", "  # result is the useful value today, but anticipate phase use in future ", "  testStatus: ", "    phase: in-progress  ", "    result: none "]}

TASK [Apply the litmus result CR] **********************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl apply -f litmus-result.yaml", "delta": "0:00:02.552845", "end": "2019-05-20 13:15:05.219952", "failed_when_result": false, "rc": 0, "start": "2019-05-20 13:15:02.667107", "stderr": "", "stderr_lines": [], "stdout": "litmusresult.litmus.io/elasticsearch-liveness configured", "stdout_lines": ["litmusresult.litmus.io/elasticsearch-liveness configured"]}

TASK [Generate the litmus result CR to reflect EOT (End of Test)] **************
skipping: [127.0.0.1] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [Analyze the cr yaml] *****************************************************
skipping: [127.0.0.1] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [Apply the litmus result CR] **********************************************
skipping: [127.0.0.1] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [Getting the application pod name] ****************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl get po -n app-esearch-ns -l app=elasticsearch -o jsonpath={.items[0].metadata.name}", "delta": "0:00:01.331136", "end": "2019-05-20 13:15:07.060016", "rc": 0, "start": "2019-05-20 13:15:05.728880", "stderr": "", "stderr_lines": [], "stdout": "elasticsearch-logging-0", "stdout_lines": ["elasticsearch-logging-0"]}

TASK [Replacing the placeholder for pod-name] **********************************
changed: [127.0.0.1] => {"changed": true, "msg": "1 replacements made"}

TASK [Replacing the placeholder for namespace] *********************************
changed: [127.0.0.1] => {"changed": true, "msg": "2 replacements made"}

TASK [Replacing the placeholder for liveness-retry-count] **********************
changed: [127.0.0.1] => {"changed": true, "msg": "1 replacements made"}

TASK [Replacing the placeholder for liveness-timeout] **************************
changed: [127.0.0.1] => {"changed": true, "msg": "1 replacements made"}

TASK [Fetch app service name] **************************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl get svc -n app-esearch-ns -o jsonpath={.items[0].metadata.name}", "delta": "0:00:01.196254", "end": "2019-05-20 13:15:10.092663", "rc": 0, "start": "2019-05-20 13:15:08.896409", "stderr": "", "stderr_lines": [], "stdout": "elasticsearch-logging", "stdout_lines": ["elasticsearch-logging"]}

TASK [Getting the port] ********************************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl get endpoints -n app-esearch-ns -o jsonpath={.items[0].subsets[0].ports[0].port}", "delta": "0:00:01.345188", "end": "2019-05-20 13:15:11.812182", "rc": 0, "start": "2019-05-20 13:15:10.466994", "stderr": "", "stderr_lines": [], "stdout": "9300", "stdout_lines": ["9300"]}

TASK [Replace the placeholder for service endpoint] ****************************
changed: [127.0.0.1] => {"changed": true, "msg": "1 replacements made"}

TASK [Replace the placeholder for service endpoint] ****************************
changed: [127.0.0.1] => {"changed": true, "msg": "1 replacements made"}

TASK [Creating elasticsearch-liveness job] *************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl create -f elasticsearch-liveness.yml -n app-esearch-ns", "delta": "0:00:01.563338", "end": "2019-05-20 13:15:14.249810", "rc": 0, "start": "2019-05-20 13:15:12.686472", "stderr": "", "stderr_lines": [], "stdout": "job.batch/elasticsearch-liveness created", "stdout_lines": ["job.batch/elasticsearch-liveness created"]}

TASK [Verifying whether liveness pod is started successfully] ******************
FAILED - RETRYING: Verifying whether liveness pod is started successfully (20 retries left).
changed: [127.0.0.1] => {"attempts": 2, "changed": true, "cmd": "kubectl get pod -n app-esearch-ns -l liveness=elasticsearch-liveness -o jsonpath={.items[0].status.phase}", "delta": "0:00:01.195519", "end": "2019-05-20 13:16:17.340889", "rc": 0, "start": "2019-05-20 13:16:16.145370", "stderr": "", "stderr_lines": [], "stdout": "Running", "stdout_lines": ["Running"]}

TASK [set_fact] ****************************************************************
ok: [127.0.0.1] => {"ansible_facts": {"flag": "Pass"}, "changed": false}

TASK [include_tasks] ***********************************************************
included: /utils/fcm/update_litmus_result_resource.yml for 127.0.0.1

TASK [Generate the litmus result CR to reflect SOT (Start of Test)] ************
skipping: [127.0.0.1] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [Analyze the cr yaml] *****************************************************
skipping: [127.0.0.1] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [Apply the litmus result CR] **********************************************
skipping: [127.0.0.1] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [Generate the litmus result CR to reflect EOT (End of Test)] **************
changed: [127.0.0.1] => {"changed": true, "checksum": "a6c4d066147b3cdb4ca64411a71b707a3a66fa51", "dest": "./litmus-result.yaml", "gid": 0, "group": "root", "md5sum": "c61b043400356460c34e0953ba0b5ed9", "mode": "0644", "owner": "root", "size": 421, "src": "/root/.ansible/tmp/ansible-tmp-1558358177.89-132558358143927/source", "state": "file", "uid": 0}

TASK [Analyze the cr yaml] *****************************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "cat litmus-result.yaml", "delta": "0:00:01.152019", "end": "2019-05-20 13:16:21.682633", "rc": 0, "start": "2019-05-20 13:16:20.530614", "stderr": "", "stderr_lines": [], "stdout": "---\napiVersion: litmus.io/v1alpha1\nkind: LitmusResult\nmetadata:\n\n  # name of the litmus testcase\n  name: elasticsearch-liveness \nspec:\n\n  # holds information on the testcase\n  testMetadata:\n    app:  \n    chaostype:  \n\n  # holds the state of testcase,  manually updated by json merge patch\n  # result is the useful value today, but anticipate phase use in future \n  testStatus: \n    phase: completed  \n    result: Pass ", "stdout_lines": ["---", "apiVersion: litmus.io/v1alpha1", "kind: LitmusResult", "metadata:", "", "  # name of the litmus testcase", "  name: elasticsearch-liveness ", "spec:", "", "  # holds information on the testcase", "  testMetadata:", "    app:  ", "    chaostype:  ", "", "  # holds the state of testcase,  manually updated by json merge patch", "  # result is the useful value today, but anticipate phase use in future ", "  testStatus: ", "    phase: completed  ", "    result: Pass "]}

TASK [Apply the litmus result CR] **********************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl apply -f litmus-result.yaml", "delta": "0:00:01.639704", "end": "2019-05-20 13:16:23.618623", "failed_when_result": false, "rc": 0, "start": "2019-05-20 13:16:21.978919", "stderr": "", "stderr_lines": [], "stdout": "litmusresult.litmus.io/elasticsearch-liveness unchanged", "stdout_lines": ["litmusresult.litmus.io/elasticsearch-liveness unchanged"]}

PLAY RECAP *********************************************************************
127.0.0.1                  : ok=21   changed=17   unreachable=0    failed=0   

```

### CODE REVIEW

File: apps/elasticsearch/liveness/test.yml [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+          shell: kubectl get endpoints -n {{ namespace }} -o jsonpath={.items[0].subsets[0].ports[0].port}
+          register: port_no
+
+        - name: Replace the placeholder for service endpoint
+          replace:
+            path: "{{ elasticsearch_liveness }}"
+            regexp: "service_name"
+            replace: "{{ name_of_service.stdout }}"
+
+        - name: Replace the placeholder for service endpoint
```
Comment (nsathyaseelan) [inline apps/elasticsearch/liveness/test.yml]: @sushma1118 This same name is used for previous task also can you change it.

Comment (sushma1118) [reply apps/elasticsearch/liveness/test.yml]: ok

Comment (sushma1118) [reply apps/elasticsearch/liveness/test.yml]: Changed the task name

File: apps/elasticsearch/liveness/test.yml [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+            replace: "{{ liveness_retry }}"   
+
+        - name: Replacing the placeholder for liveness-timeout
+          replace:
+            path: "{{ elasticsearch_liveness }}"
+            regexp: "liveness-timeout-seconds"
+            replace: "{{ liveness_timeout }}"  
+
+        - name : Fetch app service name
+          shell: kubectl get svc -n {{ namespace }} -o jsonpath={.items[0].metadata.name}
```
Comment (gprasath) [inline apps/elasticsearch/liveness/test.yml]: Use label along with namespace to find out service name.

Comment (sushma1118) [reply apps/elasticsearch/liveness/test.yml]: Done

Comment (sushma1118) [reply apps/elasticsearch/liveness/test.yml]: Used labels in test.yml and it is working fine .

Comment (sushma1118) [reply apps/elasticsearch/liveness/test.yml]: Liveness label is taken as environmental variable and is passed to test.yml

**Pull request merged**